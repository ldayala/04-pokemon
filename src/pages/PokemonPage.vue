<template>
  <h1 v-if="!pokemon">Espere por favor....</h1>
  <div v-else>
    <h1>Â¿Quien es este pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"  :vidas="vidas"/>
    <PokemonOptions :pokemons="pokemonArr" @selection="getAnswer" />
    <div v-if="showAnswer">
      <h2>{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </div>
  </div>
</template>

<script>
import PokemonOptions from "@/components/PokemonOptions.vue";
import PokemonPicture from "@/components/PokemonPicture.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: { PokemonOptions, PokemonPicture },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
      vidas:5,
      puntos:0
    };
  },
  methods: {

    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      //obtengo un nuemero aleatorio del 0-3
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },

    getAnswer(e) {
      if (this.pokemon.id == e) {
        this.message = `Correctly the pokemon's name id ${this.pokemon.name}`;
      } else {
        this.message = `Uppps the pokemon's name id ${this.pokemon.name}`;
      }
      this.showAnswer = true;
      this.showPokemon = true;
    },
    
    newGame(){
      this.pokemonArr= []
      this.pokemon= null
      this.showPokemon= false
      this.showAnswer= false
      this.message= ""
      this.mixPokemonArray()
    }

  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style>
</style>