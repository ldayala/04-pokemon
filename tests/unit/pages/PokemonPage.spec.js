import { mount, shallowMount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { mockPokemon } from "../mocks/pokemons.mock";

describe('PokemonPage Component', () => {
    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
        // ai usaramos el spy aqui arriba tambien tendriamos que hacer limpieza
    })

    test('debe hacer match con el snapshot ', () => {

        expect(wrapper.html()).toMatchSnapshot()
    });

    test('debe llamar a mixPokemonArray al montar', () => {
        //espiamos de obejot PokemoPage.methods el metodo  mixPokemonArray para ver si se llama
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        //montamos el componente para ver si llama al metodo
        shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()
    });

    test('debe hacer match con el snapshot cuando cargan los pokemons', () => {

        const wrapperw = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr:mockPokemon,
                    pokemon: mockPokemon[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                    vidas: 5,
                    puntos: 0,
                    btnValue: "Avanzar",
                }
            }
        })

        expect(wrapperw.html()).toMatchSnapshot()
    });

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper= shallowMount(PokemonPage,{
            data() {
                return {
                    pokemonArr:mockPokemon,
                    pokemon: mockPokemon[0],
                    showPokemon: true,
                    showAnswer: false,
                    message: "",
                    vidas: 5,
                    puntos: 0,
                    btnValue: "Avanzar",
                }
            }
        })
        //pokemonPicture debe existir
        expect(wrapper.find('pokemon-picture-stub').exists()).toBeTruthy()
        //PokemonOptions debe existir
        expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy()
        //pokemonPicture  attribute pokemonId===1
        expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe("1")
        //PokemonOptions attribute pokemons exista
        expect(wrapper.find('pokemon-options-stub').attributes('pokemons')).toBeTruthy()
    });

    test('pruebas con checkanswer', async() => {
        const wrapper= shallowMount(PokemonPage,{
            data() {
                return {
                    pokemonArr:mockPokemon,
                    pokemon: mockPokemon[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                    vidas: 5,
                    puntos: 0,
                    btnValue: "Avanzar",
                }
            }
        })
        //await porque se renderiuza el dom en dependencia de este metdodo llamado
         await   wrapper.vm.getAnswer(1)
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.vm.showAnswer).toBe(true)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.find('h2').text()).toBe(`Correctly the pokemon's name id ${mockPokemon[0].name}`)

        await wrapper.vm.getAnswer(3)
        expect(wrapper.vm.message).toBe(`Uppps the pokemon's name id ${mockPokemon[0].name}`)

        
    });

});