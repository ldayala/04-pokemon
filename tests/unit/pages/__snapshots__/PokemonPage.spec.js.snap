// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe hacer match con el snapshot  1`] = `<h1>Espere por favor....</h1>`;

exports[`PokemonPage Component debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>Â¿Quien es este pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false" vidas="5" puntos="0"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
